src/snowflake/connector/config_parser/parser.py

```diff
@@ -0,0 +1,298 @@
+            # If we got a TOML table we probably want it in dictionary form
+            return True, loaded_var.value
+        return True, loaded_var
+
+    def _get_config(self) -> Any:
+        """Get value from the cached config file."""
+        e = self._root_parser.conf_file_cache
+        if e is None:
+            raise ConfigParserError(
+                f"Root parser '{self._root_parser.name}' is missing file_path",
+            )
+        for k in self._nest_path[1:]:
+            e = e[k]
+        if isinstance(e, (Table, TOMLDocument)):
+            # If we got a TOML table we probably want it in dictionary form
+            return e.value
+        return e
+
+
+class ConfigParser:
+    """Read TOML configuration file with managed multi-source precedence.
+
+     This class is updatable at run-time, allowing other libraries to add their
+    own configuration options and sub-parsers. Sub-parsers allow
+     options groups to exist, e.g. the group "snowflake.cli.output" could have
+     2 options in it: debug (boolean flag) and format (a string like "json", or
+     "csv").
+
+    When a ConfigParser tries to retrieve ConfigOptions' value the _root_parser
+    will read and cache the TOML file from the file it's pointing at, afterwards
+    updating the read cache can be forced by calling read_config.
+
+    Attributes:
+        name: The name of the ConfigParser. Used for nesting and emitting
+          useful error messages.
+        file_path: Path to the file where this and all child ConfigParsers
+          should read their values out of. Can be omitted for all child
+          parsers.
+        conf_file_cache: Cache to store what we read from the TOML file.
+        _sub_parsers: List of ConfigParsers that are nested under us.
+        _options: List of ConfigOptions that are our children.
+        _root_parser: Reference to root parser. Used to efficiently propagate to
+          child options.
+        _nest_path: The names of the ConfigParsers that this parser is nested
+          under. Used to efficiently propagate to child options.
+    """
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        file_path: Path | None = None,
+    ):
+        """Create a new ConfigParser.
+
+        Args:
+            name: Name of this ConfigParser.
+            file_path: File this parser should read values from. Can be omitted
+              for all child parsers.
+        """
+        self.name = name
+        self.file_path = file_path
+        # Objects holding subparsers and options
+        self._options: dict[str, ConfigOption] = dict()
+        self._sub_parsers: dict[str, ConfigParser] = dict()
+        # Dictionary to cache read in config file
+        self.conf_file_cache: TOMLDocument | None = None
+        # Information necessary to be able to nest elements
+        #  and add options in O(1)
+        self._root_parser: ConfigParser = self
+        self._nest_path = [name]
+
+    def read_config(
```

Comment:
Does this reload from file/reconfigure the parser each time when invoked?

The comment says something about caching, but the function seems to always reload and reconfigure which probably should be the correct approach.

Maybe remove the cache part from the comment? I think it's implementation detail, and there is not really a cache mechanism in the sense that a config file will be at most read once and then cached (which could be problematic for a couple reasons).

Reply:
I've added a blurb about when I'd expect someone to call this function. Hopefully, this makes things clearer!